### ISSUE TITLE
GNMI plugin missing measurement name when when writing to influxdb

### ISSUE THREAD

**Issue by jrametta** [influxdata/telegraf#8783]
Hi, I was not seeing GNMI metric data being sent to influxdb, so I instead enabled the file output plugin with data_format="influx".  I am seeing similar error to #6391                                                             
                                                                                                                                                                                                                                    
```                                                                                                                                                                                                                                 
telegraf_1    | 2021-02-02T07:01:10Z D! [inputs.gnmi] Connection to gNMI device 10.196.21.52:50051 established                                                                                                                      
telegraf_1    | 2021-02-02T07:01:20Z D! [outputs.file] Could not serialize metric: invalid name                                                                                                                                     
telegraf_1    | 2021-02-02T07:01:20Z D! [outputs.file] Wrote batch of 1 metrics in 166.729µs                                                                                                                                        
telegraf_1    | 2021-02-02T07:01:20Z D! [outputs.file] Buffer fullness: 0 / 10000000 metrics                                                                                                                                        
```                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                    
### Relevant telegraf.conf:                                                                                                                                                                                                         
<!-- Place config in the toml code section. -->                                                                                                                                                                                     
```toml                                                                                                                                                                                                                             
  [[inputs.gnmi.subscription]]                                                                                                                                                                                                      
    origin = "device"                                                                                                                                                                                                               
    path = "/System/procsys-items/sysload-items"                                                                                                                                                                                    
    subscription_mode = "sample"                                                                                                                                                                                                    
    sample_interval = "10s"                                                                                                                                                                                                         
```                                                                                                                                                                                                                                 
If I set file output plugin to json, I can see name is empty in the output ("name":"")                                                                                                                                              
                                                                                                                                                                                                                                    
```                                                                                                                                                                                                                                 
telegraf_1    | 2021-02-02T07:03:53Z D! [inputs.gnmi] Connection to gNMI device 10.196.21.52:50051 established¬¬                                                                                                                    
telegraf_1    | {"fields":{"device:/System/procsys_items_sysload-items_loadAverage15m":"0.5","device:/System/procsys_items_sysload-items_loadAverage1m":"0.14","device:/System/procsys_items_sysload-items_loadAverage5m":"0.38","device:/System/procsys_items_sysload-items_loadAverage5sec":"9.0","device:/System/procsys_items_sysload-items_name":"sysload","device:/System/procsys_items_sysload-items_runProc":1,"device:/System/procsys_items_sysload-items_totalProc":359},"name":"","tags":{"host":"telemetry2.lab","path":"","source":"10.196.21.52"},"timestamp":1612249437}                                                                                                                     
```                                                                                                                                                                                                                                 
Per the other defect mentioned, if I configure `name_override`, then the measurement successfully written and given the configured name.                                                                                            
                                                                                                                                                                                                                                    
### System info:                                                                                                                                                                                                                    
                                                                                                                                                                                                                                    
- Telegraf 1.17.1 (git: HEAD 8907e61a)                                                                                                                                                                                              
- cisco Nexus9000 93180YC-EX                                                                                                                                                                                                        
- NXOS: version 9.3(5)                                                                                                                                                                                                              
                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                    
<!-- If your bug involves third party dependencies or services, it can be very helpful to provide a Dockerfile or docker-compose.yml that reproduces the environment you're testing against -->                                     
                                                                                                                                                                                                                                    
### Steps to reproduce:                                                                                                                                                                                                             
                                                                                                                                                                                                                                    
<!-- Describe the steps to reproduce the bug. -->                                                                                                                                                                                   
                                                                                                                                                                                                                                    
1. configure GNMI plugin with subscription to Cisco                                                                                                                                                                                 
2. start telegraf                                                                                                                                                                                                                   
                                                                                                                                                                                                                                    
### Expected behavior:                                                                                                                                                                                                              
                                                                                                                                                                                                                                    
Influxdb measurement is successfully written without error                                                                                                                                                                          
                                                                                                                                                                                                                                    
### Actual behavior:                                                                                                                                                                                                                
                                                                                                                                                                                                                                    
Receive "Could not serialize metric: invalid name"  error msg                                                                                                                                                                       
                                                                                                                                                                                                                                    
This error seems to occur for any path given, not just the one mentioned.

**Comment by jrametta** [comment]
This error is only seen when using json encoding (device capabilities reports both json and proto support)
```
[[inputs.gnmi]]
    encoding = "json"
```
After switching to proto, the problem went away. Please close this if it seems like vendor or user error.

**Comment by veorlo** [comment]
I have the same issue. It appears that the output/processor plugins used do not matter. If I dump raw metrics at the processor stage or outputs stage, they both produce an empty string for "name".

**Comment by sjwang90** [comment]
@veorlo Does your issue persist as well with just `encoding = "proto"`?

**Comment by tgupta3** [comment]
The issue still exists even if we use proto.

**Comment by sjwang90** [comment]
@sbyx I don't think this is related to #9117 but wanted to see if you had any insight on this issue too.

**Comment by veorlo** [comment]
@sjwang90 I do not experience the issue for the YANG model I'm working with while using `encoding = "proto"`. The YANG model origin is `Cisco-IOS-XR-ipv4-bgp-oper` found [here](https://github.com/YangModels/yang/blob/master/vendor/cisco/xr/731/Cisco-IOS-XR-ipv4-bgp-oper.yang). My apologies for the lacking info. I will try to add more logs/details when I can. 

That said, I've been working with Cisco to determine why some data from that YANG model is missing in the metric fields the gNMI plugin produces while using proto encoding. Using JSON, I get the missing data, but some branches within the specified path are inexplicably joined by underscores and with added list indices (e.g. `origin:/some/path/to/my/data` would return fields keyed like `origin:/some/path/to_my_0_data`, where `0` is the array index of a specific `data` list item). They have investigated and said it must be some decoding/parsing issue within the gNMI plugin. The current solution for that is to use JSON encoding because the missing data does appear at least. For now, I'm working around the field key structure. I will open a separate issue for that if I can gather sufficient evidence that would help in resolving it.

I hope this info can help despite being quite vague :)

**Comment by srebhan** [comment]
@jrametta and @veorlo can you please test a nightly build or the binary in PR #12258? Let me know if it fixes your problem!

**Comment by srebhan** [comment]
I'm closing this issue as the problem is likely fixed with PR #12258. If the problem still persists, please feel free to reopen the issue or to create a new one.
